### ISSUE TITLE
Tools jump to non-existent points of interest in light scene

### ISSUE THREAD

**Issue by Nancy-Salpepi** [phetsims/geometric-optics#426]
**Test device**
MacBook Air (m1 chip)

**Operating System**
12.3.1

**Browser**
Chrome

**Problem description**
For https://github.com/phetsims/qa/issues/796 and related to https://github.com/phetsims/geometric-optics/issues/395

In lens screen, with the light and the concave lens: using the 'J' key to jump to points of interest takes the rulers and position markers to non-existent points of interest. It seems like they are moving to where the virtual image and second point of virtual image would be if it was a framed object. 

**Steps to reproduce**
1. Select the concave lens in the Lens screen
2. Select the light
3. Tab to the horizontal ruler and press space to select
4. Press the 'J' key several times 
5. Repeat steps 3-4 with the vertical ruler and position markers

**Visuals**

https://user-images.githubusercontent.com/87318828/163054309-beff300b-dfd4-45a4-b742-9d38cbbe589d.mp4

<details>
Troubleshooting information:
!!!!! DO NOT EDIT !!!!!
Name: ‪Geometric Optics‬
URL: https://phet-dev.colorado.edu/html/geometric-optics/1.1.1-rc.1/phet/geometric-optics_all_phet.html
Version: 1.1.1-rc.1 2022-04-11 14:16:55 UTC
Features missing: applicationcache, applicationcache, touch
Flags: pixelRatioScaling
User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.75 Safari/537.36
Language: en-US
Window: 1430x690
Pixel Ratio: 2/1
WebGL: WebGL 1.0 (OpenGL ES 2.0 Chromium)
GLSL: WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor: WebKit (WebKit WebGL)
Vertex: attribs: 16 varying: 31 uniform: 1024
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 32)
Max viewport: 16384x16384
OES_texture_float: true
Dependencies JSON: {}
</details>

**Comment by arouinfar** [comment]
Really nice find @Nancy-Salpepi.

I only see this happing with the light/concave lens combo. The markers also jump to the same x-coordinates as the horizontal ruler, but reveal a non-zero y-coordinate. I have no idea what these points of interest are meant to represent, but if you add the second light and move them around, the marker always touches the central ray.

<img width="637" alt="image" src="https://user-images.githubusercontent.com/8419308/163056373-3d9a7e5d-e4ed-4bad-bcdc-7150847ca562.png">
<img width="493" alt="image" src="https://user-images.githubusercontent.com/8419308/163056461-d68e84b1-6fb0-4b94-a671-a805175e7b90.png">

As weird as this looks, I really would like to punt on it until 1.2. It is critical to publish the maintenance release to fix #424 as soon as possible because it is a very visible bug that users are going to notice.

**Comment by pixelzoom** [comment]
This "jump" feature is so complicated and fragile.  I hope someone finds it useful, but I'll be surprised if most people can figure out what it does without scaffolding.  That said...

Agreed, let's defer this to 1.2 release.

**Comment by pixelzoom** [comment]
The problem occurs with all tools.  As @Nancy-Salpepi surmised, these are the optical image points, where an optical image would be formed.  No optical image is formed in the Light scene, but these points are nevertheless interesting -- sometimes.  When the optic is converging (convex lens) these points are were the rays converge, on the right side of the lens, and they are useful measurement points -- see screenshot below.  When the optic is diverging (concave lens), the rays do not converge, and these points are NOT useful.

I'll investigate ignoring these points for the Light scene when the lens is concave.

<img width="1029" alt="screenshot_1651" src="https://user-images.githubusercontent.com/3046552/163193779-655d0016-f520-4deb-8aa4-19562f7f0214.png">

**Comment by pixelzoom** [comment]
I also realized that the positions of the light spots are not really "interesting" for the concave lens. The diameter of the spot is huge, so there's no distint "spot" on the screen.  And the position doesn't aligned with the center marginal ray (see screenshot below) -- I'm not sure if that's correct or incorrect, and it's not easy to determine. 

So... I've also exlcuded the positions of the light spots for the concave lens.

<img width="1503" alt="screenshot_1652" src="https://user-images.githubusercontent.com/3046552/163217276-0a907a4a-8c83-4662-b2eb-68961d8d00ef.png">

**Comment by pixelzoom** [comment]
Another problem, related to convex lens.  There's currently nothing that knows whether the lights spots intersect the project screen.  The light spots are always drawn, and simply clipped to the projections screen.  So if the center of the light spot is not on the projection screen, jumping to the light spot can look like the 2 examples below.  This is the position of the light spot _in the vertical plane of the screen_.

@arouinfar do you think this needs to be addressed?

<img width="1034" alt="screenshot_1654" src="https://user-images.githubusercontent.com/3046552/163219603-4f36e9df-9d33-4877-9937-b00edafbcdc1.png">

<img width="1035" alt="screenshot_1653" src="https://user-images.githubusercontent.com/3046552/163219626-9c2a82ef-6c34-483c-9bb8-a63dadbd92da.png">

**Comment by pixelzoom** [comment]
@arouinfar while you're replying.... Please review the changes that I've made so far.  For concave lens in the Light scene, I've removed (1) the optical image position and (2) the light spot positions from the jump points.

**Comment by pixelzoom** [comment]
I thought of a way to address https://github.com/phetsims/geometric-optics/issues/426#issuecomment-1098213935.  It involves a significant amount of work, but nothing particularly tricky:

* Add a constant to ProjectionScreen that indicates how tall the screen is at its center-x. ` const averageScreenHeight` looks like it's what we need, change it to `this.averageScreenHeight`.

* Add `visibleProperty` to LightSpot, derived `postionProperty`, `diameterProperty`, the projection screen's `positionProperty`, the projection screen's height its center-x point, and the opticalImage `visibleProperty` (for animation).  Something like this in LightSpot.ts:

```js
this.visibleProperty = new DerivedPropery( 
  [ this.positionProperty, this.diameterProperty, projectionScreen.positionProperty, opticalImage.visibleProperty ],
  ( postion, diameter, projectionScreenPosition, opticalImageVisible ) => 
    opticalImageVisible &&
    position.y >= projectionScreenPosition.y - projectionScreen.height / 2 - diameter / 2 &&
    position.y < = projectionScreenPosition.y + projectionScreen.height / 2 + diameter / 2,  {
...
} );
 
* Change `visibleProperty` for each LightSpotNode to be simply be something like this in LightSpotNode.ts:

```js
visibleProperty: lightSpot1.visibleProperty
``` 

* Make a change something like this in LightSceneNode:

```js
-      visibleProperty: DerivedProperty.and( [ lightPropagationEnabledProperty, scene.opticalImage1.visibleProperty ], {
+      visibleProperty: DerivedProperty.and( [ lightPropagationEnabledProperty, lightSpot1.visibleProperty ], {
...
-      visibleProperty: DerivedProperty.and(
        [ lightPropagationEnabledProperty, scene.opticalImage2.visibleProperty, visibleProperties.secondPointVisibleProperty ], {
+      visibleProperty: DerivedProperty.and(
        [ lightPropagationEnabledProperty, lightSpot2.visibleProperty, visibleProperties.secondPointVisibleProperty ], {
```

**Comment by pixelzoom** [comment]
Well, that went smoother than expected...  I did not exactly follow the plan outlined in https://github.com/phetsims/geometric-optics/issues/426#issuecomment-1098398015.  Instead of LightSpot `visibleProperty`, I added `intersectsProjectionScreenProperty`.   If any part of a light spot intersects the projection screen, then the associated LightSpotNode will be visible, and the light spot's `positionProperty` (its center point) will be a jump point for tools.

@arouinfar ready for review in master.

**Comment by arouinfar** [comment]
Thanks @pixelzoom. Here are the notable changes:

1. Concave lens: optical image positions are skipped
2. Concave lens: light spot position is skipped 
3. Convex lens: light spot skipped if fully off-screen
4. Convex lens: if light spot is fully off-screen, the optical image position is also skipped

It's not clear if (4) is intentional or not. In general, the optical image isn't particularly interesting for the Light scene. I am fine with skipping it altogether. It seems odd to jump there only when the lens is convex and the light spot is visible on the screen.

**Comment by pixelzoom** [comment]
@arouinfar and I met on Zoom, and decided to delete the optical image positions from the Light scene jump points. Done in the above commit, and we both verified.  Labeling as fixed-awaiting-deploy until the next 1.2 RC.

**Comment by KatieWoe** [comment]
Things are looking good in dev.1
